keps/sig-api-machinery/3775-cel-mutating-admission/README.md

```diff
@@ -0,0 +1,638 @@
+
+The first policy contains two expressions that mutate Pods. The first expression sets pod.spec.dnsPolicy
+field to None and the second defaults pod.spec.dnsConfig to a config pointing to a custom nameserver:
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: MutatingAdmissionPolicy
+metadata:
+ name: "demo-policy.example.com"
+spec:
+ matchConstraints:
+   resourceRules:
+   - apiGroups:   [""]
+     apiVersions: ["v1"]
+     operations:  ["CREATE", "UPDATE"]
+     resources:   ["pods"]
+ mutations:
+   - expression: 'JSONPatch[{ op: "replace", path: "/spec/dnsPolicy", value: "None" }]'
+     condition: "object.spec.dnsPolicy != "None"
+   - expression: 'JSONPatch[{ op: "replace", path: "/spec/dnsConfig", value: { "nameservers": ["1.2.3.4"], "searches": ["ns1.svc.cluster-domain.example", "my.dns.search.suffix"] } }]'
+```
+
+Here is another example of a policy that defaults all Deployments to have 5 replicas:
+```yaml
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: MutatingAdmissionPolicy
+metadata:
+ name: "demo-policy.example.com"
+spec:
+ matchConstraints:
+   resourceRules:
+   - apiGroups:   ["apps"]
+     apiVersions: ["v1"]
+     operations:  ["CREATE", "UPDATE"]
+     resources:   ["deployments"]
+ mutations:
+   - expression: 'JSONPatch[{ op: "replace", path: "/spec/replicas", value: 5 }]'
+```
+
+Lastly, here is an example policy that sets the imagePullPolicy for all containers to “Always”:
+```yaml
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: MutatingAdmissionPolicy
+metadata:
+ name: "demo-policy.example.com"
+spec:
+ matchConstraints:
+   resourceRules:
+   - apiGroups:   ["apps"]
+     apiVersions: ["v1"]
+     operations:  ["CREATE", "UPDATE"]
+     resources:   ["deployments"]
+ mutations:
+   - expression: '[1…size(object.spec.containers)].map(i, JSONPatch[{op: “replace”, path: /spec/containers/i/imagePullPolicy, value: “Always”}])'
+```
+
+NOTE: the expression `[1...size(object.spec.containers)]` is not supported by CEL yet but there are on-going discussions to add
+a new macro and type to support this use-case.
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a cluster admin, I would like to have a mutating admission policy that adds a HTTP_PROXY
+and HTTPS_PROXY environment variable to every container in my cluster. Deploying a mutating
+admission webhook seems cumbersome since the proxy URL is static.
+
+#### Story 2
+
+As a cluster admin, I would like to have a mutating admission policy that ensures all
+`ValidatingWebhookConfiguration` and `MutatingWebhookConfiguration` objects exclude
+`leases` and all objects in the `kube-system` namespace. This is currently not possible
+today since `MutatingWebhookConfiguration` cannot mutate `ValidatingWebhookConfiguration` or
+`MutatingWebhookConfiguration` objects.
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
```

Comment:
What is K8s stance on the issue that this risks turning the API Server into a Turing machine?

Mutation of one field can cause mutation of a different field to either be enabled or not.

For example: imagine the following object:

`{"a": 1, "b": 2, "c": 3}`

And the following mutations (note that the if statements make the individual mutations idempotent):

Mutation X:
If `"a" < "b"`, swap `"a"` and `"b"`

Mutation Y:
If `"b"  < "c"`, swap `"b"` and `"c"`

Mutation Z:
If `"a" > "c"`, swap `"a"` and `"c"`

The output (assuming application order X -> Y -> Z)  will bounce between:

`{"a": 1, "b": 3, "c": 2}`
`{"a": 1, "b": 2, "c": 3}`

By the third iteration, we have come full-circle and the cycle repeats.

Note that each mutator is idempotent in isolation, but the set of mutators together is not.

Other cross-mutator failure modes could include growing a value until its infinitely large (e.g. infinitely prepending to a string).

We can make sure each individual API server request terminates by limiting the number of mutation iterations, but there is a larger control loop in the form of controllers, which the API server cannot affect directly.

Are we okay with having the same risk profile with this syntax as we do with mutation webhooks?

My hunch is that there is a probabilistic nature to these failure modes... the more mutations you have, the more likely one is to occur. Right now, because writing a mutation requires building a webhook (which is a significant effort) and because there is an effective cap on the # of mutation webhooks per request due to latency concerns, the number of mutations in the wild is fairly low, this might change that.

I'd also expect some of these issues to have a delayed occurrence: as acceptance/usage increases, this is when bad interactions become more likely, so I'm not sure how observable this problem would be in, say, the alpha stage.

Because mutators are also running in the API server directly, the impact of a vulnerability may be more severe.

The solution isn't obvious to me (there are ways to define sets of mutator primitives that are guaranteed to be well-behaved, but then their functionality is necessarily limited). But having a plan to deal with or detect these edge cases seems prudent.

Reply:
> Are we okay with having the same risk profile with this syntax as we do with mutation webhooks?

One limitation with webhooks is that the API server knows nothing about what fields the webhook might mutate until is sends an admission review response. With CEL, it seems a lot more plausible we could have a design where the apiserver to be able to inspect the MutatingAdmissionPolicy to know which fields the policy intends to mutate.

Also knowing which fields a MutatingAdmissionPolicy "looks at" when performing a mutation could help?  This might be do-able. We take a CEL expression, compile it, traverse the AST and build up an understanding which fields the expression accesses.

Say we did all that and we are able to effectively have a set of fields that each mutating policy reads, and a set of fields that the policy mutates.

Would that be useful? Can we then throw that data into some routines that would allow us to do things like find optimal orderings of the mutations (for DAGs) and detect and ban cycles?

Even if we don't built all that out in the near term, it might be useful to have this sketched out so that we can avoid design decisions that would prevent us from building this out later?

Reply:
I think that could help, for sure, though reasoning about whether a structure can create loops is tricky business, so if we wanted to go that route, we'd need to put some time into analysis.

I think the key thing here that may make your suggestion "convergent" (a term I coined for a system that eventually settles on a stable result through recursion and therefore is idempotent for future applications), is that the fields tested are different from the fields being read.

My intuition is ordering will wind up not mattering for convergence, but will be useful for things like "which mutator wins" (e.g. mutate image pull policy AFTER adding sidecards).

The other half of this is practicality... will it be the case that adding restrictions like this preclude some commonly requested use cases? Does it add an effective cap to the total number of mutations?

Then there is the matter of adding a new mutator may render pre-existing mutators in-conflict, presumably disabling them which could have interesting operational effects.

I don't think perfection is a possibility, TBH, but worth thinking through what the tradeoffs are and making them intentionally.

Reply:
Let's continue to discuss.

At the very least, it sounds like we should it a desirable design goal to be able to know which fields a mutator "reads" and "writes". What exactly we would do with that is still up in the air, but is sound like we keep our options open if we have access to that information.